src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,142 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            Node<T> newNode = new Node<>(null, value, null);
+            head = tail = newNode;
+            head.next = tail;
+            tail.prev = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0 && isEmpty()) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            checkIndex(index);
+            Node<T> node = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(node.prev, value, node);
+            node.prev.next = newNode;
+            node.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findNodeByIndex(index).item;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private int findNodeByValue(T value) {
+        Node<T> current = head;
+        for (int i = 0; i < size(); i++) {
+            if (Objects.equals(current.item, value)) {
+                return i;
+            }
+            current = current.next;
+        }
+        return -1;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.item;
+        currentNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
```

Comment:
Please implement this comment:

![image](https://github.com/mate-academy/jv-linked-list/assets/140905715/3b873ddc-2c76-4e37-96d7-f2f78b3c991f)